pandas/io/formats/csvs.py

```diff
@@ -36,8 +36,20 @@ def __init__(self, obj, path_or_buf=None, sep=",", na_rep='',
         if path_or_buf is None:
             path_or_buf = StringIO()
 
+        self._compression_arg = compression
```

Comment:
didn't you change get_filepath_or_buffer to already handle this? why is this special cased here?

Reply:
I did not modify `get_filepath_or_buffer`, though I suppose I certainly could to support taking a `dict` as `compression`. The `self._compression_arg` is there to avoid changing `self.compression` from only ever holding the inferred compression method, while `self._compression_arg` would include any additional arguments if `compression` was passed as a `dict`. I do now think that I should have instead had a `self.compression_args` hold this dict with the `method` key popped.

Reply:
@jreback With regard to `get_filepath_or_buffer` (and similarly `_infer_compression`), I've added a function `_get_compression_method` which handles the case where `compression` is given as a `dict` to `to_csv`. It extracts the compression method string before passing to `get_filepath_buffer` or `_infer_compression`. 
Would it be preferable then to keep both functions' original functionality where `compression` may only be a string/None, or include a call to `_get_compression_method` in both to handle dicts?